File: QadeerApp.Web/Modules/Membership/Account/SignUp/SignUpForm.cs
using QadeerApp.Administration;

namespace QadeerApp.Membership;

[FormScript("Membership.SignUp")]
public class SignUpForm
{
    [Required(true), Placeholder("full name")]
    public string DisplayName { get; set; }
    [LookupEditor(typeof(DepartmentRow), InplaceAdd = false)]
    [Required(true)]
    public int DepartmentId { get; set; }
    [LookupEditor(typeof(SpecializationRow), CascadeFrom = "DepartmentId", CascadeField = "DepartmentId", InplaceAdd = false)]
    public int? SpecializationId { get; set; }
    [EmailAddressEditor, Required(true), Placeholder("email")]
    public string Email { get; set; }
    [EmailAddressEditor, Required(true), Placeholder("confirm email")]
    public string ConfirmEmail { get; set; }
    [PasswordEditor, Required(true), Placeholder("password")]
    public string Password { get; set; }
    [PasswordEditor, Required(true), Placeholder("confirm password")]
    public string ConfirmPassword { get; set; }
}

File: QadeerApp.Web/Modules/ServerTypes/Membership/SignUpForm.ts
import { EmailAddressEditor, initFormType, LookupEditor, PasswordEditor, PrefixedContext, StringEditor } from "@serenity-is/corelib";

export interface SignUpForm {
    DisplayName: StringEditor;
    DepartmentId: LookupEditor;
    SpecializationId: LookupEditor;
    Email: EmailAddressEditor;
    ConfirmEmail: EmailAddressEditor;
    Password: PasswordEditor;
    ConfirmPassword: PasswordEditor;
}

export class SignUpForm extends PrefixedContext {
    static readonly formKey = 'Membership.SignUp';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!SignUpForm.init) {
            SignUpForm.init = true;

            var w0 = StringEditor;
            var w1 = LookupEditor;
            var w2 = EmailAddressEditor;
            var w3 = PasswordEditor;

            initFormType(SignUpForm, [
                'DisplayName', w0,
                'DepartmentId', w1,
                'SpecializationId', w1,
                'Email', w2,
                'ConfirmEmail', w2,
                'Password', w3,
                'ConfirmPassword', w3
            ]);
        }
    }
}
